packages/nitro-protocol/contracts/AssetHolder.sol

```diff
@@ -312,14 +333,138 @@ contract AssetHolder is IAssetHolder {
+        bytes32 destination
+    ) internal {
+        Outcome.AllocationItem[] memory allocation = abi.decode(
+            allocationBytes,
+            (Outcome.AllocationItem[])
+        );
+        uint256 balance = holdings[guarantorChannelId];
+        uint256 affordsForDestination;
+        uint256 residualAllocationAmount;
+        uint256 i; // indexes target allocations
+
+        for (uint256 j = 0; j < guarantee.destinations.length; j++) {
+            if (balance == 0) {
+                revert('_claim | guarantorChannel affords 0 for destination');
+            }
+            // for each destination in the guarantee,
+            // find the first corresponding allocationItem in the target allocation
+            bytes32 guaranteeDestination = guarantee.destinations[j];
+            for (i = 0; i < allocation.length; i++) {
+                if (allocation[i].destination == guaranteeDestination) {
+                    // decrease balance
+                    uint256 _amount = allocation[i].amount;
+                    if (balance < _amount) {
+                        if (guaranteeDestination == destination) {
+                            affordsForDestination = balance;
+                            residualAllocationAmount = _amount - balance;
+                            break;
+                            // i will point to index that should be modified or removed in the target outcome
+                        }
+                        balance = 0; // this isn't used after we break
+                    } else {
+                        if (guaranteeDestination == destination) {
+                            affordsForDestination = _amount;
+                            residualAllocationAmount = 0;
+                            break;
+                            // i will point to index that should be modified or removed in the target outcome
+                        }
+                        balance = balance.sub(_amount); // this isn't used after we break
+                    }
+                    break;
+                }
+            }
+            if (affordsForDestination > 0) {
+                // stop lopping as soon as we found the destination in both outcomes such that we can pay something out
+                break;
+            }
+        }
+
+        require(affordsForDestination > 0, '_claim | guarantor affords 0 for destination');
+
+        // effects
+        holdings[guarantorChannelId] -= affordsForDestination;
+
+        // construct new outcome for target
+        if (residualAllocationAmount > 0) {
+            // new allocation identical save for a single entry
+            Outcome.AllocationItem[] memory newAllocation = new Outcome.AllocationItem[](
+                allocation.length
+            );
+            for (uint256 k = 0; k < allocation.length; k++) {
+                newAllocation[k] = allocation[k];
+                if (allocation[k].destination == destination) {
```

Comment:
Could you use `k==i` here instead?

Reply:
Yes, and that is actually better since here we can potentially go around again and find a second (repeated) entry if it exists (whereas we are not doing that above).

Reply:
20bfad5